doc/incudine.1
.TH INCUDINE 1 "January 2023" "" ""
.SH NAME
incudine - Music/DSP programming environment for Common Lisp.
.SH SYNOPSIS
\fBincudine\fR [\fIOPTIONS\fR] [\fIFILE1\fR [\fIFILE2\fR ...]]
         [\fIOPTIONS\fR] [\fIFILE1\fR [\fIFILE2\fR ...]] ...
         [\fIUSER-OPTIONS\fR]
.SH DESCRIPTION
Standalone executable for Incudine, a Music/DSP programming
environment for Common Lisp.

Incudine is useful to design software synthesizers or sound plugins
from scratch, exploiting the expressive power of Common Lisp, without
the constraint to use pre-built unit generators. It is also a
compositional tool that allows to produce high quality sounds
controllable at the sample level, defining and redefining the digital
signal processors and the musical structures on-the-fly.

Incudine introduces the Virtual UGen (VUG), a new mechanism for
defining primitive unit generators by writing concise and reusable
code. The definition of a VUG contains lisp code and/or other nested
VUGs (or "real" UGens), and it is re-arranged and compiled efficiently
during the creation of a DSP.

Incudine works with SBCL, an implementation of ANSI Common Lisp with a
high-performance native compiler.
.SH OPTIONS
.TP
\fB\-b\fR, \fB\-\-sndfile\-buffer\-size\fR <\fIint\fR>
Buffer size to read/write a soundfile.
.TP
\fB\-B\fR, \fB\-\-block\-size\fR <\fIframes\fR>
Block size for block\-by\-block processing.
.TP
\fB\-c\fR, \fB\-\-channels\fR <\fIint\fR>
Number of the output channels.
.TP
\fB\-\-client\-name\fR <\fIname\fR>
Name of the client for the audio server.
.TP
\fB\-\-compile\-score\-contents\fR
Compile the contents of the score.
.TP
\fB\-F\fR, \fB\-\-data\-format\fR <\fItype\fR>
Format of the sample for the output file.
.TP
\fB\-\-data\-format\-list\fR
Available audio formats.
.TP
\fB\-\-default\-table\-size\fR <\fIint\fR>
Default size of a table for an oscillator.
.TP
\fB\-d\fR, \fB\-\-duration\fR <\fIseconds\fR>
Duration of the output file.
.TP
\fB\-\-debug\fR
Print debug info.
.TP
\fB\-\-disk\-guard\-size\fR <\fIseconds\fR>
Size of the output file with undefined duration.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print a summary of the available options on the standard output.
.TP
\fB\-H\fR, \fB\-\-header\-type\fR <\fItype\fR>
Type of the header for the output file.
.TP
\fB\-\-header\-type\-list\fR
Available header types.
.TP
\fB\-i\fR, \fB\-\-infile\fR <\fIfilename\fR>
Sound input filename or `\-' for standard input.
.TP
\fB\-\-input\-channels\fR <\fIint\fR>
Number of the input channels.
.TP
\fB\-\-interpret\-score\-contents\fR
Interpret the contents of the score (default).
.TP
\fB\-L\fR, \fB\-\-logfile\fR <\fIfilename\fR>
Logging file.
.TP
\fB\-\-lisp\-version\fR
Print version information of SBCL and exit.
.TP
\fB\-\-logtime\fR (\fBsec\fR|\fBsamp\fR)
Log message with time in seconds or in samples.
.TP
\fB\-\-max\-number\-of\-channels\fR <\fIint\fR>
Max number of the channels.
.TP
\fB\-\-max\-number\-of\-nodes\fR <\fIint\fR>
Max number of the nodes.
.TP
\fB\-N\fR, \fB\-\-no\-realtime\fR
Stop realtime.
.TP
\fB\-\-nrt\-edf\-heap\-size\fR <\fIint\fR>
Max number of the events in non realtime.
.TP
\fB\-\-nrt\-pool\-size\fR <\fIbytes\fR>
Size of the pool for temporary C malloc in nrt.
.TP
\fB\-\-nrt\-priority\fR <\fIint\fR>
Priority of the non\-realtime thread.
.TP
\fB\-o\fR, \fB\-\-outfile\fR <\fIfilename\fR>
Sound output filename or `\-' for standard output.
.TP
\fB\-p\fR, \fB\-\-period\fR <\fIint\fR>
Frames per buffer (used only with PortAudio).
.TP
\fB\-\-pad\fR <\fIseconds\fR>
Extend the duration of the output file.
.TP
\fB\-r\fR, \fB\-\-rate\fR <\fIint\fR>
Sample rate.
.TP
\fB\-R\fR, \fB\-\-realtime\fR
Start realtime with the previous options.
.TP
\fB\-\-receiver\-priority\fR <\fIint\fR>
Priority of the thread for a receiver.
.TP
\fB\-\-rt\-edf\-heap\-size\fR <\fIint\fR>
Heap size for realtime scheduling.
.TP
\fB\-\-rt\-pool\-size\fR <\fIbytes\fR>
Size of the pool for the C heap used in realtime.
.TP
\fB\-\-rt\-priority\fR <\fIint\fR>
Priority of the realtime thread.
.TP
\fB\-s\fR <\fIfilename\fR>
Process a score file or `\-' for standard input.

In non-realtime, without the option \fB\-o\fR or \fB\-\-outfile\fR, the
name of the produced sound file is derived from the name of the score file
(if standard input, the output will be sent to standard output).
.TP
\fB\-\-sample\-pool\-size\fR <\fIbytes\fR>
Size of the pool for the C arrays defined in DSP!
.TP
\fB\-\-score\-args\fR <\fIlist\fR>
Score function arguments.
.TP
\fB\-\-sound\-velocity\fR <\fIreal\fR>
Velocity of the sound at 22°C, 1 atmosfera.
.TP
\fB\-\-swank\-server\fR <\fIport\fR>
Start Swank server.
.TP
\fB\-T\fR, \fB\-\-tempo\fR <\fIbpm\fR>
Initial tempo in beats per minute.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
More verbose.
.TP
\fB\-\-version\fR
Print version information and exit.
.SS Metadata
.HP
\fB\-\-title\fR <\fItitle\fR>
.HP
\fB\-\-copyright\fR <\fIcopyright\fR>
.HP
\fB\-\-software\fR <\fIsoftware\fR>
.HP
\fB\-\-artist\fR <\fIartist\fR>
.HP
\fB\-\-comment\fR <\fIcomment\fR>
.HP
\fB\-\-date\fR <\fIdate\fR>
.HP
\fB\-\-album\fR <\fIalbum\fR>
.HP
\fB\-\-license\fR <\fIlicense\fR>
.HP
\fB\-\-tracknumber\fR <\fItrack\fR>
.HP
\fB\-\-genre\fR <\fIgenre\fR>
.SS SBCL options
.TP
\fB\-\-sysinit\fR <\fIfilename\fR>
System\-wide init\-file to use instead of default.
.TP
\fB\-\-userinit\fR <\fIfilename\fR>
Per\-user init\-file to use instead of default.
.TP
\fB\-\-no\-sysinit\fR
Inhibit processing of any system\-wide init\-file.
.TP
\fB\-\-no\-userinit\fR
Inhibit processing of any per\-user init\-file.
.TP
\fB\-\-disable\-debugger\fR
Invoke sb\-ext:disable\-debugger.
.TP
\fB\-\-noinform\fR
Suppress the printing of the banner at startup.
.TP
\fB\-\-noprint\fR
Run a Read\-Eval Loop without printing results.
.TP
\fB\-\-script\fR [<\fIfilename\fR>]
Skip #! line, disable debugger, avoid verbosity.
.TP
\fB\-\-quit\fR
Exit with code 0 after option processing.
.TP
\fB\-\-non\-interactive\fR
Sets both \fB\-\-quit\fR and \fB\-\-disable\-debugger\fR.
.TP
\fB\-\-eval\fR <\fIform\fR>
Form to eval when processing this option.
.TP
\fB\-\-load\fR <\fIfilename\fR>
File to load when processing this option.
.SH EXAMPLES
.nf
cat >foo.cudo <<'---'
(dsp! simple (freq db (env envelope) gate)
  (stereo (* (envelope env gate 1 #'free)
             (sine freq (db->linear db) 0))))
---

cat >quux.rego <<'---'
with (env1 (or (first score-args) (make-adsr .1 .09 .9 .5)))

0   simple  440  -14  env1  1
0   simple  220  -20  env1  1
0.5 simple  225  -26  env1  1
1   simple  448  -14  env1  1
3   simple  450  -20  (make-adsr 1 .2 .7 1 :curve :sin)  1
5   set-control 0 :gate 0
---

incudine foo.cudo -v -s quux.rego

incudine foo.cudo --pad 1.2 -s quux.rego

incudine foo.cudo --score-args "(list (make-perc .01 3))" \\
                  --pad 1.2 -s quux.rego

incudine foo.cudo -o "battimenti zincati.wav" -s quux.rego

echo '0 simple 1000 -6 (make-perc .1 1.5) 1' \\
     | incudine foo.cudo -d 2 -s - > rego-from-stdin-1.wav

cat quux.rego | incudine foo.cudo -o rego-from-stdin-2.wav -s -

incudine foo.cudo --pad 1.2 -o - -H ircam -F float \\
         -s quux.rego | sox - quux.wv

incudine foo.cudo -d 6.5 -s quux.rego

incudine foo.cudo --pad 1.2 -H flac -s quux.rego

incudine foo.cudo --pad 1.2 -H ogg -F vorbis -s quux.rego

incudine foo.cudo --pad 1.2 --title "Passame n'elettrodo" \\
         --artist "Fabro Cadabro Band" -s quux.rego

.SH FILES
~/.incudinerc is the configuration file.
.SH SEE ALSO
Tutorials in HTML and text formats in incudine/doc/ directory.

Incudine home page at http://incudine.sourceforge.net
.SH BUGS
Report bugs at incudine-devel@lists.sourceforge.net
.SH AUTHORS
Incudine was written by Tito Latini.


contrib/editors/emacs/incudine.el
;;; incudine.el --- major mode for editing Incudine sources

;; Copyright (c) 2013-2025 Tito Latini

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
;; incudine.el

;; Installation:

;; Add the following lines to your `.emacs' file
;;
;;   (push "/path/to/incudine/contrib/editors/emacs" load-path)
;;   (require 'incudine)

;;; Code:
(require 'slime)

(defgroup incudine nil
  "Major mode for editing incudine code."
  :group 'languages)

(defvar incudine-mode-hook nil
  "Hook called when a buffer enters Incudine mode.")

(defvar incudine-rego-mode-hook nil
  "Hook called when a buffer enters Incudine Rego mode.")

(defcustom incudine-scratch-message nil
  "Initial message displayed in *incudine-scratch* buffer.
If this is nil, no message will be displayed."
  :type '(choice (text :tag "Message")
                 (const :tag "none" nil))
  :group 'incudine)

(defcustom incudine-save-buffer-before-play t
  "Save the modified rego file before playback."
  :type 'boolean
  :group 'incudine)

(defun incudine-buffer-name (string)
  (concat "*incudine-" string "*"))

(defvar incudine-scratch-buffer (incudine-buffer-name "scratch"))

(defvar incudine-rego-files-walk nil
  "Association list where the CAR of each element is the buffer related
to an included rego file and CDR is the buffer of the parent.
Used during meta-dotting from a rego file to an included rego file and
vice versa.")

(defun incudine-scratch ()
  "Switch to Incudine scratch buffer."
  (interactive)
  (let ((buffer (get-buffer incudine-scratch-buffer)))
    (unless buffer
      (setq buffer (get-buffer-create incudine-scratch-buffer))
      (with-current-buffer buffer
        (incudine-mode)
        (slime-mode t)
        (if incudine-scratch-message
            (insert incudine-scratch-message))))
    (switch-to-buffer buffer)))

(defun incudine-eval (string &rest args)
  (let ((current-prefix-arg nil))
    (slime-interactive-eval
      (if args (apply #'format string args) string))))

(defun incudine-eval-defun ()
  "Evaluate the current toplevel form."
  (interactive)
  (slime-eval-defun))

(defun incudine-show-repl ()
  "Show REPL in other window."
  (interactive)
  (let ((buffer (or (and (fboundp 'slime-repl-buffer)
                         ;; Ignore if killed buffer.
                         (funcall 'slime-repl-buffer))
                    (let ((proc (get-process "inferior-lisp")))
                      (and proc (process-buffer proc))))))
    (when buffer
      (switch-to-buffer-other-window buffer))))

(defun incudine-repl-clear-buffer ()
  "Clear the REPL buffer."
  (interactive)
  (let* ((buffer (and (fboundp 'slime-repl-buffer)
                      (funcall 'slime-repl-buffer)))
         (proc (unless buffer (get-process "inferior-lisp"))))
    (when (or buffer proc)
      (save-window-excursion
        (with-current-buffer (or buffer (process-buffer proc))
          (cond (buffer
                 (funcall 'slime-repl-insert-prompt)
                 (funcall 'slime-repl-clear-buffer))
                (t (comint-clear))))))))

(defun incudine-prev-defun (&optional n)
  "Jump at the end of the previous defun.

If N is non-NIL, do it that many times."
  (interactive "p")
  (end-of-defun)
  (beginning-of-defun)
  (if n
      (dotimes (i n) (beginning-of-defun))
      (beginning-of-defun))
  (forward-sexp))

(defun incudine-next-defun (&optional n)
  "Jump at the end of the next defun.

If N is non-NIL, do it that many times."
  (interactive "p")
  (end-of-defun)
  (if n
      (dotimes (i n) (end-of-defun))
      (forward-sexp))
  (beginning-of-defun)
  (forward-sexp))

(defun incudine-eval-and-next-fn ()
  "Eval the function and jump to the next defun."
  (interactive)
  (slime-eval-defun)
  (incudine-next-defun))

(defun incudine-eval-and-prev-fn ()
  "Eval the function and jump to the previous defun."
  (interactive)
  (slime-eval-defun)
  (incudine-prev-defun))

(defun incudine-eval-print-last-expression (string &optional save-point-p)
  (interactive (list (slime-last-expression) t))
  (if save-point-p
      ;; A version of `slime-eval-print' with `save-excursion'.
      (slime-eval-async `(swank:eval-and-grab-output ,string)
        (lambda (result)
          (save-excursion
            (cl-destructuring-bind (output value) result
              (insert "\n" output value)))))
      (progn
        (insert "\n")
        (slime-eval-print string))))

(defun prefix-numeric-value0 (n)
  (if n (prefix-numeric-value n) 0))

(defun incudine-rt-start (&optional block-size)
  "Start the real-time thread.

If BLOCK-SIZE is positive, set the new block size before starting."
  (interactive "P")
  (let ((n (prefix-numeric-value0 block-size)))
    (if (> n 0)
        (incudine-eval
          "(progn (incudine:set-rt-block-size %d) (values (incudine:rt-start) (incudine:block-size)))"
          n)
        (incudine-eval "(incudine:rt-start)"))))

(defun incudine-rt-stop ()
  "Stop the real-time thread."
  (interactive)
  (incudine-eval "(incudine:rt-stop)"))

(defun incudine-free-node (&optional id)
  "Stop to play a node of the graph.

If ID is negative, call `incudine:stop' instead of `incudine:free'.

If ID is zero, call `incudine:flush-pending' before `incudine:free'."
  (interactive "P")
  (let ((n (prefix-numeric-value0 id)))
    (incudine-eval (cond ((= n 0) "(progn (incudine:flush-pending) (incudine:free 0))")
                         ((< n 0) "(incudine:stop %d)")
                         (t "(incudine:free %d)"))
                   (abs n))))

(defun incudine-pause-node (&optional id)
  "Pause node.

If ID is non-NIL, pause the node with identifier ID."
  (interactive "P")
  (incudine-eval "(incudine:pause %d)"
                 (prefix-numeric-value0 id)))

(defun incudine-unpause-node (&optional id)
  "Unpause node.

If ID is non-NIL, unpause the node with identifier ID."
  (interactive "P")
  (incudine-eval "(incudine:unpause %d)"
                 (prefix-numeric-value0 id)))

(defvar incudine-node-tree-buffer (incudine-buffer-name "node-tree"))

(defvar incudine-node-tree-dump-command nil)

(defvar incudine-node-tree-map
  (let ((map (make-sparse-keymap)))
    (define-key map "g" 'incudine-redump-graph)
    map))

(defun incudine-dump-graph (&optional id)
  "Print informations about the graph of nodes.

If ID is non-NIL, print info about the group with identifier ID."
  (interactive "P")
  (let ((form (format "(incudine:dump (incudine:node %d) *standard-output*)"
                      (prefix-numeric-value0 id))))
    (setq incudine-node-tree-dump-command form)
    (slime-eval-async `(swank:pprint-eval ,form)
      (lambda (result)
        (slime-with-popup-buffer
            (incudine-node-tree-buffer
             :mode (when (and (boundp 'text-scale-mode)
                              (symbol-value 'text-scale-mode))
                     ;; Keep text-scale minor mode.
                     'text-scale-mode))
          (insert result)
          (use-local-map incudine-node-tree-map)
          (kill-line -1))))))

(defun incudine-redump-graph ()
  (interactive)
  (slime-eval-async `(swank:pprint-eval ,incudine-node-tree-dump-command)
    (lambda (result)
      (with-current-buffer incudine-node-tree-buffer
        (let ((inhibit-read-only t))
          (erase-buffer)
          (insert result)
          (kill-line -1)
          (goto-char (point-min)))))))

(defun incudine-live-nodes ()
  "Print the number of the live nodes."
  (interactive)
  (incudine-eval "(incudine:live-nodes)"))

(defun incudine-scheduled-events ()
  "Print the number of the scheduled events."
  (interactive)
  (incudine-eval
    "(incudine.util:rt-eval (:return-value-p t) (incudine.edf:heap-count))"))

(defun incudine-gc ()
  "Initiate a garbage collection."
  (interactive)
  (incudine-eval "(incudine.util::gc :full t)"))

(defun incudine-bytes-consed-in (&optional time)
  "Rough estimate of the bytes consed in TIME seconds."
  (interactive "p")
  (incudine-eval "(incudine.util:get-bytes-consed-in %d)"
                 (prefix-numeric-value time)))

(defun incudine-rt-memory-free-size ()
  "Display the free realtime memory."
  (interactive)
  (incudine-eval
   "(values (incudine.util:get-foreign-sample-free-size)
            (incudine.util:get-rt-memory-free-size))"))

(defun incudine-peak-info (&optional ch)
  "Display the peak info of the channel CH (0 by default).

If CH is a negative number, reset the meters."
  (interactive "P")
  (let ((value (prefix-numeric-value0 ch)))
    (if (< value 0)
        (incudine-eval "(incudine:reset-peak-meters)")
        (incudine-eval "(incudine:peak-info %d)" value))))

(defun incudine-xruns-info (&optional reset-p)
  "Display the number of the occurred xruns and the time in samples of
the last xrun.

If RESET-P is non-NIL, set the number of xruns to zero."
  (interactive "P")
  (incudine-eval "(incudine:rt-xruns %S)" (and reset-p t)))

(defun incudine-set-logger-level (value)
  (incudine-eval "(setf (incudine.util:logger-level) %s)" value))

(defun incudine-logger-level-choice (c)
  "Set Logger Level from a single character."
  (interactive "cLogger level? (e)rror, (w)arn, (i)nfo or (d)ebug")
  (when (member c '(?e ?w ?i ?d))
    (incudine-set-logger-level
      (pcase c
        (?e ":ERROR")
        (?w ":WARN")
        (?i ":INFO")
        (?d ":DEBUG")))))

(defun incudine-set-logger-stream (value)
  (incudine-eval "(setf incudine.util:*logger-stream* %s)" value))

(defun incudine-logger-stream-choice (c)
  "Set Logger Stream from a single character."
  (interactive "cLogger stream? (e)rror-output, (s)tandard-output or (d)ebug-io")
  (when (member c '(?e ?s ?d))
    (incudine-set-logger-stream
      (pcase c
        (?e "CL:*ERROR-OUTPUT*")
        (?s "CL:*STANDARD-OUTPUT*")
        (?d "CL:*DEBUG-IO*")))))

(defun incudine-set-logger-time (value)
  (incudine-eval "(setf (incudine.util:logger-time) %s)" value))

(defun incudine-logger-time-choice (c)
  "Set Logger Time from a single character."
  (interactive "cLogger time? (S)amples, (s)econds or (n)il")
  (when (member c '(?S ?s ?n))
    (incudine-set-logger-time
      (pcase c
        (?n "NIL")
        (?s ":SEC")
        (?S ":SAMP")))))

(defun incudine-slime-macroexpand (form)
  (slime-eval-async `(cl:second (swank:eval-and-grab-output ,form))
    #'slime-initialize-macroexpansion-buffer))

(defun incudine-dsp/ugen-expand ()
  "If the form at point starts with DSP! or DEFINE-UGEN, prompt the
arguments and display the generated code."
  (interactive)
  (let ((string (slime-sexp-at-point)))
    (when (string-match "^[ \t]*(\\([^ \t:]+:\\)?\\(dsp!\\|define-ugen\\) " string)
      (let* ((type (if (char-equal (aref string (1+ (match-beginning 2))) ?s)
                       "dsp"
                       "ugen"))
             (args (read-from-minibuffer (concat type " args: "))))
        (incudine-slime-macroexpand
          (format "(funcall (incudine.vug:%s-debug %s %s)"
                  type (substring string (match-end 0)) args))))))

(defun incudine-buffer-string ()
  (buffer-substring-no-properties (buffer-end -1) (buffer-end 1)))

(defun incudine-play-buffer ()
  (let ((slime-log-events nil))
    (incudine-eval
      (format "(progn (incudine:rt-start) (funcall (incudine:regostring->function %S)))"
              (incudine-buffer-string)))))

(defun incudine-play-regofile ()
  "Eval the edited rego file and schedule the obtained events."
  (interactive)
  (let ((file-name (buffer-file-name)))
    (if (not file-name)
        (incudine-play-buffer)
      (when (and incudine-save-buffer-before-play (buffer-modified-p))
        (save-buffer))
      (incudine-eval
        "(progn (incudine:rt-start) (funcall (incudine:regofile->function %S)))"
        file-name))))

(defun incudine-regobuffer-to-function ()
  (let ((slime-log-events nil))
    (incudine-eval
      (format "(incudine:regostring->function %S)" (incudine-buffer-string)))))

(defun incudine-regofile-to-function (&optional name)
  "Define the function to interpret the edited rego file.
With a `C-u' prefix, prompt the function name.

The score statement ~:score-function-name:~ or ~:score-local-function-name:~
is an alternative method to set the function name."
  (interactive (list (and current-prefix-arg
                          (read-from-minibuffer "function name: "
                                                nil nil t nil "nil"))))
  (let ((file-name (buffer-file-name)))
    (if (not file-name)
        (incudine-regobuffer-to-function)
      (when (and incudine-save-buffer-before-play (buffer-modified-p))
        (save-buffer))
      (incudine-eval "(incudine:regofile->function %S '%S)" file-name name))))

(defun incudine-fix-rego-files-walk ()
  "Remove the killed buffers from incudine-rego-files-walk."
  (setq incudine-rego-files-walk
        (cl-delete-if-not (lambda (x)
                            (and (buffer-live-p (car x))
                                 (buffer-live-p (cdr x))))
                          incudine-rego-files-walk)))

(defun incudine-find-name (&optional name where)
  "If the current line is an `include' statement, edit the included
file name, otherwise edit a lisp definition or call `find-tag'."
  (interactive (list (and (not current-prefix-arg) (slime-symbol-at-point))))
  (unless (save-excursion
            (beginning-of-line)
            (when (re-search-forward
                    "^[; \t]*include +\"\\(\\(\\\\\"\\|[^\"]\\)+\\)\""
                    (point-at-eol) t)
              (let ((parent (current-buffer)))
                (incudine-fix-rego-files-walk)
                (find-file (match-string-no-properties 1))
                (cl-pushnew (cons (current-buffer) parent)
                            incudine-rego-files-walk :key 'car))))
    (let ((name (or name (slime-read-symbol-name "Edit Definition of: "))))
      (if (and (slime-connected-p) (slime-find-definitions name))
          (slime-edit-definition name where)
          (find-tag name)))))

(defun incudine-find-name-retract ()
  "If we are editing a rego file, goto the location of the parent
rego file or call `tags-loop-continue'."
  (interactive)
  (let* ((from (current-buffer))
         (entry (assoc from incudine-rego-files-walk)))
    (if entry
        (let ((to (cdr entry)))
          (incudine-fix-rego-files-walk)
          (when (buffer-live-p to)
            (setq incudine-rego-files-walk
                  (delete entry incudine-rego-files-walk))
            (switch-to-buffer to)))
        (tags-loop-continue))))

(defun incudine-regofile-to-sexp ()
  "Display the expansion of the edited rego file."
  (interactive)
  (let ((file-name (buffer-file-name)))
    (when (and file-name (buffer-modified-p))
      (save-buffer))
    (incudine-slime-macroexpand
      (if file-name
          (format "(incudine:regofile->sexp %S)" file-name)
          (format "(incudine::regostring->sexp %S)" (incudine-buffer-string))))))

(defun incudine-regofile-to-list ()
  "Display the list of events obtained from the edited rego file."
  (interactive)
  (let ((file-name (buffer-file-name)))
    (when (and file-name (buffer-modified-p))
      (save-buffer))
    (incudine-slime-macroexpand
      (if file-name
          (format "(incudine:regofile->list %S)" (buffer-file-name))
          (format "(incudine:regostring->list %S)" (incudine-buffer-string))))))

(defun incudine-regofile-to-scheduled-events ()
  "Display the scheduled events obtained from the edited rego file."
  (interactive)
  (let ((file-name (buffer-file-name)))
    (when (and file-name (buffer-modified-p))
      (save-buffer))
    (let ((form-beg
            "with-schedule ,@(mapcar (lambda (x) `(at (* ,(car x) *sample-rate*) #',(cadr x) ,@(cddr x)))"))
      (incudine-slime-macroexpand
        (if file-name
            (format "`(%s (incudine:regofile->list %S)))"
                    form-beg (buffer-file-name))
            (format "`(%s (incudine:regostring->list %S)))"
                    form-beg (incudine-buffer-string)))))))

(font-lock-add-keywords
  'incudine-mode
  '((;; define-vug define-vug-macro define-ugen define-ugen-control-setter
     ;; defscore-statement defun* defmacro* lambda* dsp!
     "\\(d\\(?:ef\\(?:ine-\\(?:ugen\\(?:-control-setter\\)?\\|vug\\(?:-macro\\)?\\)\\|macro\\*\\|score-statement\\|un\\*\\)\\|sp!\\)\\|lambda\\*\\)\\_>\\s *\\(\\(?:\\sw\\|\\s_\\|\\\\.\\)+\\)?"
     (1 font-lock-keyword-face nil t)
     (2 font-lock-function-name-face nil t))
    ("(\\(vuglet\\)\\_>"
     (1 font-lock-keyword-face nil t))))

(put 'define-vug 'common-lisp-indent-function '(4 &lambda &body))
(put 'define-ugen 'common-lisp-indent-function '(4 4 &lambda &body))
(put 'dsp! 'common-lisp-indent-function '(4 &lambda &body))
(put 'vuglet 'common-lisp-indent-function
     '((&whole 4 &rest (&whole 1 &lambda &body)) &body))
(put 'initialize 'common-lisp-indent-function 0)

(defun incudine-mode-common-map (map)
  (define-key map [C-M-return] 'incudine-free-node)
  (define-key map "\C-cv" 'incudine-show-repl)
  (define-key map "\C-cs" 'incudine-scratch)
  (define-key map "\C-c\M-o" 'incudine-repl-clear-buffer)
  (define-key map "\C-crs" 'incudine-rt-start)
  (define-key map "\C-crq" 'incudine-rt-stop)
  (define-key map "\C-cp" 'incudine-pause-node)
  (define-key map "\C-cu" 'incudine-unpause-node)
  (define-key map "\C-cgc" 'incudine-gc)
  (define-key map "\C-cgb" 'incudine-bytes-consed-in)
  (define-key map "\C-cid" 'incudine-dsp/ugen-expand)
  (define-key map "\C-cie" 'incudine-scheduled-events)
  (define-key map "\C-cig" 'incudine-dump-graph)
  (define-key map "\C-cim" 'incudine-rt-memory-free-size)
  (define-key map "\C-cin" 'incudine-live-nodes)
  (define-key map "\C-cip" 'incudine-peak-info)
  (define-key map "\C-cix" 'incudine-xruns-info)
  (define-key map "\C-cll" 'incudine-logger-level-choice)
  (define-key map "\C-cls" 'incudine-logger-stream-choice)
  (define-key map "\C-clt" 'incudine-logger-time-choice))

(defvar incudine-mode-map
  (let ((map (make-sparse-keymap "Incudine")))
    (incudine-mode-common-map map)
    (define-key map [C-return] 'incudine-eval-and-next-fn)
    (define-key map [C-S-return] 'incudine-eval-and-prev-fn)
    (define-key map [M-return] 'incudine-eval-defun)
    (define-key map [prior] 'incudine-prev-defun)
    (define-key map [next] 'incudine-next-defun)
    map)
  "Keymap for Incudine mode.")

(defvar incudine-scratch-mode-map
  (let ((map (make-sparse-keymap)))
    (set-keymap-parent map incudine-mode-map)
    map))

(slime-define-keys incudine-scratch-mode-map
  ("\C-j" 'slime-eval-print-last-expression)
  ((kbd "C-S-j") 'incudine-eval-print-last-expression))

(defvar incudine-rego-mode-map
  (let ((map (make-sparse-keymap "Incudine Rego")))
    (incudine-mode-common-map map)
    (define-key map [f5] 'incudine-regofile-to-function)
    (define-key map [f9] 'incudine-play-regofile)
    (define-key map "\C-cRe" 'incudine-regofile-to-scheduled-events)
    (define-key map "\C-cRl" 'incudine-regofile-to-list)
    (define-key map "\C-cRs" 'incudine-regofile-to-sexp)
    (define-key map "\M-." 'incudine-find-name)
    (define-key map "\M-," 'incudine-find-name-retract)
    map)
  "Keymap for Incudine Rego mode.")

(defvar incudine-mode-menu-list
  (list "Incudine"
        (list "REPL"
              ["Show REPL" incudine-show-repl t]
              ["REPL Clear Buffer" incudine-repl-clear-buffer t])
        (list "Realtime"
              ["RT Start" incudine-rt-start t]
              ["RT Stop" incudine-rt-stop t]
              ["Number of Scheduled Events" incudine-scheduled-events t]
              ["Peak Info"
               (incudine-peak-info
                 (string-to-number (read-from-minibuffer "Channel: " "0")))
               :keys "C-c i p"]
              ["Xruns Info" incudine-xruns-info t]
              ["Reset Peak Meters" (incudine-peak-info -1) t]
              ["Reset Xruns" (incudine-xruns-info t) t])
        (list "Graph"
              ["Stop Playing" incudine-free-node t]
              ["Pause" incudine-pause-node t]
              ["Unpause" incudine-unpause-node t]
              ["Number of Live Nodes" incudine-live-nodes t]
              ["Print Graph" incudine-dump-graph t])
        (list "Memory"
              ["Garbage Collection" incudine-gc t]
              ["RT Memory Free Size" incudine-rt-memory-free-size t])
        (list "Debugging"
              ["DSP/UGEN Generated Code" incudine-dsp/ugen-expand t])
        (list "Logger"
              ["Log Level" incudine-logger-level-choice t]
              ["Log Stream" incudine-logger-stream-choice t]
              ["Log Time"  incudine-logger-time-choice t])
        ["Scratch buffer" incudine-scratch t]))

(easy-menu-define incudine-mode-menu incudine-mode-map
  "Menu used in Incudine mode."
  incudine-mode-menu-list)

(easy-menu-define incudine-rego-mode-menu incudine-rego-mode-map
  "Menu used in Incudine Rego mode."
  (let* ((l (copy-sequence incudine-mode-menu-list))
         (rt-entry (assoc "Realtime" (cdr l)))
         (debug-entry (assoc "Debugging" (cdr l))))
    (setf (nthcdr 3 rt-entry)
          (cons ["Play Rego File" incudine-play-regofile t]
                (nthcdr 3 rt-entry)))
    (setf (cdr debug-entry)
          (list ["Rego File to List of Events" incudine-regofile-to-list t]
                ["Rego File to Scheduled Events"
                 incudine-regofile-to-scheduled-events t]
                ["Rego File to Sexp" incudine-regofile-to-sexp t]))
    l))

(defvar incudine-rego-font-lock-keywords
  '(("^;.*$" . font-lock-comment-face)))

(add-to-list 'auto-mode-alist '("\\.\\(cudo\\|rexi\\)$" . incudine-mode))
(add-to-list 'auto-mode-alist '("\\.rego$" . incudine-rego-mode))

(define-derived-mode incudine-mode lisp-mode "Incudine"
  "Major mode for incudine.

\\{incudine-mode-map}"
  (use-local-map
    (if (buffer-file-name)
        incudine-mode-map
        incudine-scratch-mode-map))
  (easy-menu-add incudine-mode-menu)
  (run-hooks 'incudine-mode-hook))

(define-derived-mode incudine-rego-mode org-mode "Incudine Rego"
    "Major mode for incudine rego.

\\{incudine-rego-mode-map}"
  (use-local-map incudine-rego-mode-map)
  (easy-menu-add incudine-rego-mode-menu)
  (push '(";.*$" . font-lock-comment-face) org-font-lock-keywords)
  (setq-local comment-start ";")
  (setq-local comment-end "")
  (setq-local comment-start-skip ";+ *")
  (setq-local comment-add 1)
  (setq-local comment-column 40)
  (setq-local comment-region-function 'comment-region-default)
  (setq-local uncomment-region-function 'uncomment-region-default)
  (when (boundp 'org-emphasis-alist)
    ;; No emphasized text between the symbols //
    (setq-local org-emphasis-alist
      (assoc-delete-all "/" (copy-alist org-emphasis-alist))))
  (run-hooks 'incudine-rego-mode-hook))

(provide 'incudine)
